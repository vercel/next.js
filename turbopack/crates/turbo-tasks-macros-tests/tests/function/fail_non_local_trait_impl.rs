#![feature(arbitrary_self_types)]
#![feature(arbitrary_self_types_pointers)]
#![allow(dead_code)]

use turbo_tasks::Vc;

#[turbo_tasks::value]
struct ExampleStruct;

#[turbo_tasks::value(transparent, local)]
struct IntegersVec(Vec<Vc<u32>>);

#[turbo_tasks::value_trait]
trait ExampleTrait {
    fn return_contains_unresolved_vc(self: Vc<Self>) -> Vc<IntegersVec>;
}

#[turbo_tasks::value_impl]
impl ExampleTrait for ExampleStruct {
    #[turbo_tasks::function(non_local_return)]
    fn return_contains_unresolved_vc(self: Vc<Self>) -> Vc<IntegersVec> {
        Vc::cell(Vec::new())
    }
}

fn main() {}
