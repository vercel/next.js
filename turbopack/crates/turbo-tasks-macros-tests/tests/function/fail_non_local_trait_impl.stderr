error[E0277]: the trait bound `IntegersVec: NonLocalValue` is not satisfied
  --> tests/function/fail_non_local_trait_impl.rs:20:29
   |
20 |     #[turbo_tasks::function(non_local_return)]
   |                             ^^^^^^^^^^^^^^^^ the trait `NonLocalValue` is not implemented for `IntegersVec`
   |
   = help: the following other types implement trait `NonLocalValue`:
             &T
             &mut T
             ()
             (A, Z, Y, X, W, V, U, T)
             (B, A, Z, Y, X, W, V, U, T)
             (C, B, A, Z, Y, X, W, V, U, T)
             (D, C, B, A, Z, Y, X, W, V, U, T)
             (E, D, C, B, A, Z, Y, X, W, V, U, T)
           and $N others
note: required by a bound in `assert_returns_non_local_value`
  --> $WORKSPACE/turbopack/crates/turbo-tasks/src/macro_helpers.rs
   |
   | pub fn assert_returns_non_local_value<ReturnType, Rv>()
   |        ------------------------------ required by a bound in this function
...
   |     Rv: NonLocalValue + Send,
   |         ^^^^^^^^^^^^^ required by this bound in `assert_returns_non_local_value`
